File: packages/agent/src/infrastructure/events/types/state.ts
/**
 * @fileoverview Session State and Related Types
 *
 * Types for session state reconstruction, messages, and queries.
 */

import type { EventId, SessionId, WorkspaceId, BranchId } from './branded.js';
import type { EventType } from './base.js';
import type { ContentBlock } from './message.js';
import type { TokenUsage } from './token-usage.js';
import type { SessionEvent } from './union.js';

// =============================================================================
// Event Creation Helpers
// =============================================================================

export type CreateEventInput<T extends SessionEvent> = Omit<T, 'id' | 'timestamp' | 'sequence' | 'checksum'>;

// =============================================================================
// Message Type (for API compatibility)
// =============================================================================

/** Message format for API calls */
export interface Message {
  role: 'user' | 'assistant' | 'system' | 'toolResult';
  content: string | ContentBlock[];
  /** For toolResult messages: the ID of the tool call this result corresponds to */
  toolCallId?: string;
  /** For toolResult messages: whether the result is an error */
  isError?: boolean;
}

/**
 * Message paired with its associated event IDs.
 *
 * Consolidates the parallel arrays pattern where messages[] and messageEventIds[]
 * had to stay in sync. A message can have multiple eventIds when consecutive
 * messages of the same role are merged (for proper deletion tracking).
 *
 * The eventIds array may be empty for:
 * - Synthetic messages (tool results, compaction summaries)
 * - Messages created during the current session (not yet persisted)
 */
export interface MessageWithEventId {
  message: Message;
  /** Event IDs associated with this message. Multiple IDs when messages are merged. */
  eventIds: (string | undefined)[];
}

// =============================================================================
// Session State (Reconstructed from Events)
// =============================================================================

export interface SessionState {
  /** Session ID */
  sessionId: SessionId;
  /** Workspace ID */
  workspaceId: WorkspaceId;
  /** The head event this state is at */
  headEventId: EventId;
  /** Current model */
  model: string;
  /** Working directory */
  workingDirectory: string;
  /** Messages with their associated event IDs (unified, no parallel arrays) */
  messagesWithEventIds: MessageWithEventId[];
  /** Total token usage */
  tokenUsage: TokenUsage;
  /** Turn count */
  turnCount: number;
  /** Current provider */
  provider?: string;
  /** System prompt */
  systemPrompt?: string;
  /** Current reasoning level (for extended thinking models) */
  reasoningLevel?: 'low' | 'medium' | 'high' | 'xhigh';
  /** Current turn number (deprecated, use turnCount) */
  currentTurn?: number;
  /** Session metadata */
  metadata?: SessionMetadata;
  /** Active files being worked on */
  activeFiles?: string[];
  /** Whether session is archived (derived from archived_at) */
  isArchived?: boolean;
  /** Branch information */
  branch?: {
    id: BranchId;
    name: string;
  };
  /** Timestamp of this state */
  timestamp?: string;
}

export interface SessionMetadata {
  title?: string;
  tags: string[];
  created: string;
  lastActivity: string;
  forkedFrom?: {
    sessionId: SessionId;
    eventId: EventId;
  };
  custom: Record<string, unknown>;
}

// =============================================================================
// Tree Structures
// =============================================================================

export interface Branch {
  /** Branch identifier */
  id: BranchId;
  /** Human-readable name */
  name: string;
  /** Session this branch belongs to */
  sessionId: SessionId;
  /** Root event of this branch (fork point) */
  rootEventId: EventId;
  /** Current head event */
  headEventId: EventId;
  /** Number of events in branch */
  eventCount: number;
  /** Creation timestamp */
  created: string;
  /** Last activity */
  lastActivity: string;
  /** Is this the main/default branch */
  isDefault: boolean;
}

export interface TreeNode {
  /** Event at this node */
  eventId: EventId;
  /** Event type */
  type: EventType;
  /** Timestamp */
  timestamp: string;
  /** Summary for display */
  summary: string;
  /** Child nodes (branches from this point) */
  children: TreeNode[];
  /** Depth in tree */
  depth: number;
  /** Is this the current head of any branch */
  isHead: boolean;
  /** Is this a branching point (multiple children) */
  isBranchPoint: boolean;
  /** Branch this node belongs to */
  branchId?: BranchId;
}

export interface TreeNodeCompact {
  id: EventId;
  parentId: EventId | null;
  type: EventType;
  timestamp: string;
  summary: string;
  hasChildren: boolean;
  childCount: number;
  depth: number;
  isBranchPoint: boolean;
  isHead: boolean;
}

// =============================================================================
// Query Types
// =============================================================================

export interface SearchResult {
  eventId: EventId;
  sessionId: SessionId;
  type: EventType;
  timestamp: string;
  /** Matched content snippet */
  snippet: string;
  /** Match relevance score */
  score: number;
}

export interface SessionSummary {
  sessionId: SessionId;
  workspaceId: WorkspaceId;
  title?: string;
  eventCount: number;
  messageCount: number;
  branchCount: number;
  tokenUsage: TokenUsage;
  created: string;
  lastActivity: string;
  /** Whether session is archived (derived from archived_at IS NOT NULL) */
  isArchived: boolean;
  tags: string[];
}

// =============================================================================
// Workspace
// =============================================================================

export interface Workspace {
  id: WorkspaceId;
  path: string;
  name?: string;
  created: string;
  lastActivity: string;
  sessionCount: number;
}

File: packages/agent/crates/tron-events/src/types/state.rs
//! State types for sessions, workspaces, and search results.
//!
//! These match the TypeScript `state.ts` types for wire compatibility.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::generated::EventType;
use super::payloads::TokenUsage;

/// A reconstructed message from the event history.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    /// Message role.
    pub role: String,
    /// Message content (string for user/system, array for assistant).
    pub content: Value,
    /// Tool call ID (for `toolResult` messages).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    /// Whether this is an error result.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_error: Option<bool>,
}

/// A message with its source event IDs.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageWithEventId {
    /// The reconstructed message.
    pub message: Message,
    /// Source event IDs (multiple when messages are merged).
    pub event_ids: Vec<Option<String>>,
}

/// Full state of a session at a given point.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionState {
    /// Session ID.
    pub session_id: String,
    /// Workspace ID.
    pub workspace_id: String,
    /// Head event ID.
    pub head_event_id: String,
    /// Current model.
    pub model: String,
    /// Working directory.
    pub working_directory: String,
    /// Reconstructed messages.
    pub messages_with_event_ids: Vec<MessageWithEventId>,
    /// Aggregate token usage.
    pub token_usage: TokenUsage,
    /// Number of completed turns.
    pub turn_count: i64,
    /// Provider name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// Current system prompt.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_prompt: Option<String>,
    /// Reasoning level.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_level: Option<String>,
    /// Session metadata.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<SessionMetadata>,
    /// Whether the session has ended.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_ended: Option<bool>,
    /// Branch info.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<BranchRef>,
    /// Timestamp of the state.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
}

/// Session metadata.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionMetadata {
    /// Session title.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Tags.
    #[serde(default)]
    pub tags: Vec<String>,
    /// Creation timestamp.
    pub created: String,
    /// Last activity timestamp.
    pub last_activity: String,
    /// Fork source.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub forked_from: Option<ForkRef>,
    /// Custom metadata.
    #[serde(default)]
    pub custom: Value,
}

/// Fork reference in session metadata.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForkRef {
    /// Source session ID.
    pub session_id: String,
    /// Source event ID.
    pub event_id: String,
}

/// Branch reference.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct BranchRef {
    /// Branch ID.
    pub id: String,
    /// Branch name.
    pub name: String,
}

/// Branch state.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Branch {
    /// Branch ID.
    pub id: String,
    /// Branch name.
    pub name: String,
    /// Session ID.
    pub session_id: String,
    /// Root event ID.
    pub root_event_id: String,
    /// Head event ID.
    pub head_event_id: String,
    /// Event count.
    pub event_count: i64,
    /// Creation timestamp.
    pub created: String,
    /// Last activity timestamp.
    pub last_activity: String,
    /// Whether this is the default branch.
    pub is_default: bool,
}

/// Session summary (lightweight, for list views).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSummary {
    /// Session ID.
    pub session_id: String,
    /// Workspace ID.
    pub workspace_id: String,
    /// Session title.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Event count.
    pub event_count: i64,
    /// Message count.
    pub message_count: i64,
    /// Branch count.
    pub branch_count: i64,
    /// Aggregate token usage.
    pub token_usage: TokenUsage,
    /// Creation timestamp.
    pub created: String,
    /// Last activity timestamp.
    pub last_activity: String,
    /// Whether the session has ended.
    pub is_ended: bool,
    /// Tags.
    #[serde(default)]
    pub tags: Vec<String>,
}

/// Workspace info.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Workspace {
    /// Workspace ID.
    pub id: String,
    /// Absolute path.
    pub path: String,
    /// Display name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Creation timestamp.
    pub created: String,
    /// Last activity timestamp.
    pub last_activity: String,
    /// Number of sessions.
    pub session_count: i64,
}

/// Search result from FTS5 search.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchResult {
    /// Event ID.
    pub event_id: String,
    /// Session ID.
    pub session_id: String,
    /// Event type.
    #[serde(rename = "type")]
    pub event_type: EventType,
    /// Timestamp.
    pub timestamp: String,
    /// Highlighted snippet.
    pub snippet: String,
    /// BM25 relevance score.
    pub score: f64,
}
